### PR TITLE
Websocket intergration and tls

### LINKED ISSUE #15: Implement a WebSocket Client from libsignal

### PR DESCRIPTION
**Pull request by MJHC**
This pull request adds:
- tls
- http -> https redirect
- websocket skeleton
- test certifactions and tls cert generator script
closes #15

### CODE REVIEW

File: server/src/server.rs [hunk 1]
```diff
@@ -97,6 +112,37 @@ async fn handle_put_registration<T: SignalDatabase>(
...
 
+// redirect from http to https
+async fn redirect_http_to_https(addr: SocketAddr, http: u16, https: u16) {
+    fn make_https(host: String, uri: Uri, http: u16, https: u16) -> Result<Uri, BoxError> {
+        let mut parts = uri.into_parts();
+
+        parts.scheme = Some(axum::http::uri::Scheme::HTTPS);
+
+        if parts.path_and_query.is_none() {
+            parts.path_and_query = Some("/".parse().unwrap());
```
Comment (Mikkel-Boje) [inline server/src/server.rs:123]: Please do error handling

File: server/src/server.rs [hunk 2]
```diff
@@ -97,6 +112,37 @@ async fn handle_put_registration<T: SignalDatabase>(
...
+    }
+
+    let redirect = move |Host(host): Host, uri: Uri| async move {
+        match make_https(host, uri, http, https) {
+            Ok(uri) => Ok(Redirect::permanent(&uri.to_string())),
+            Err(_) => Err(StatusCode::BAD_REQUEST),
+        }
+    };
+
+    let listener = tokio::net::TcpListener::bind(addr).await.unwrap();
```
Comment (Mikkel-Boje) [inline server/src/server.rs:139]: Error handling

File: server/src/server.rs [hunk 3]
```diff
@@ -97,6 +112,37 @@ async fn handle_put_registration<T: SignalDatabase>(
...
+
+    axum::serve(listener, redirect.into_make_service())
+        .await
+        .unwrap();
```
Comment (Mikkel-Boje) [inline server/src/server.rs:143]: Error handling

File: server/src/server.rs [hunk 4]
```diff
@@ -240,12 +286,67 @@ async fn post_link_device_endpoint(State(state): State<SignalServerState<Postgre
...
+    mut socket: WebSocket,
+    who: SocketAddr,
+) {
+    /* authenticated_device should be put into the socket_manager,
+    we should probably have a representation like in the real server */
+    state.socket_manager.add_ws(who, socket).await;
+
+    while let Some(msg_res) = state.socket_manager.ws_recv(&who).await {
+        let msg = match msg_res {
+            Ok(x) => x,
```
Comment (Mikkel-Boje) [inline server/src/server.rs:319]: Please name x and y

File: server/src/server.rs [hunk 5]
```diff
@@ -257,8 +358,11 @@ pub async fn start_server() -> Result<(), Box<dyn std::error::Error>> {
         .max_age(Duration::from_secs(5184000))
         .allow_credentials(true)
         .allow_headers([AUTHORIZATION, CONTENT_TYPE, CONTENT_LENGTH, ACCEPT, ORIGIN]);
+
     let state = SignalServerState::<PostgresDatabase>::new().await;
+
     let app = Router::new()
+        .route("/", get(|| async { "Hello from Signal Server" }))
```
Comment (Mikkel-Boje) [inline server/src/server.rs:365]: What is the intention behind this?

File: server/src/socket.rs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    pub async fn get_ws(&self, who: &SocketAddr) -> Option<Arc<Mutex<WebSocket>>> {
+        self.sockets.read().await.get(who).cloned()
+    }
+
+    // do not use this, unless is socket_handler
+    pub async fn ws_recv(&self, who: &SocketAddr) -> Option<Result<Message, axum::Error>> {
+        match self.get_ws(who).await {
+            Some(x) => x.lock().await.recv().await,
+            None => None,
```
Comment (Mikkel-Boje) [inline server/src/socket.rs:55]: Can we do something like this instead?
```suggestion
        self.get_ws(who).await.and_then(|ws| ws.lock.await.recv().await)
```

File: server/src/socket.rs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    // only use this
+    pub async fn ws_send(
+        &self,
+        who: &SocketAddr,
+        message: Message,
+    ) -> Option<Result<(), axum::Error>> {
+        match self.get_ws(who).await {
+            Some(x) => Some(x.lock().await.send(message).await),
```
Comment (Mikkel-Boje) [inline server/src/socket.rs:66]: Same here

**Review by Mikkel-Boje — CHANGES_REQUESTED** [review]

File: server/src/server.rs (hunk 5 above)
Comment (MJHC) [reply server/src/server.rs:339]: you can get on it and see that it works

Comment (Mikkel-Boje) [reply server/src/server.rs:339]: Sure

File: server/src/socket.rs (hunk 6 above)
Comment (MJHC) [reply server/src/socket.rs:55]: this fails

**Review by Mikkel-Boje — APPROVED** [review]
Super :+1:

**Review by Laursen79 — APPROVED** [review]

**Pull request merged**